metrics: lookup podman container names

jelly opened the pull request:
As follow up of showing podman containers in the CPU usage overview, map
container ids to names. The cgroup holds no information on the name of
the podman container, so to look up the name we have to shell out to the
podman command line utility. As shelling out to podman is expensive we
only look up podman container mappings when a container makes it in top
cpu usage list. Multiple users can run a podman containers on a system
so we look up the userid based on the cgroup path and use sudo to
execute podman.

jelly commented:
There might be some UX issues left for when no mapping is found, we probably want to show something, maybe?

Or when you are in limited access and click on a root container you'll see nothing. We could map the curent user == containerid and then show something.

jelly commented on pkg/metrics/metrics.jsx:
```diff
@@ -391,37 +399,62 @@ class CurrentMetrics extends React.Component {
             if (is_container) {
-                cockpit.jump("/podman");
+                if (podNameInCache(podNameMapping, uid, name)) {
+                    cockpit.jump("/podman#/?name=" + podName(podNameMapping, uid, name));
```
This would show a link for users which can not be viewed in cockpit-podman for example:

I have a user named `jelle` who runs a container, with administrative access I can see their cgroups but cockpit-podman only shows your containers + administrative access containers, so the links don't really work.

As we know the userid, we could check if the uid is 0 or current logged and then show a link. Otherwise show for who the container owner is (that would require a username lookup) and ofcourse has to be done in cgroupRow.

martinpitt replied:
I'd actually prefer to not show these at all. If you still want to, I don't veto this, though -- then for that I like your suggestion.

jelly replied:
I'm ok with leaving it out, having a follow up which explains what `user@1000` is, so ideally it would be: `user session for jelle` but maybe that's too long already.

martinpitt replied:
How about "user jelle", similar to "pod foo"? But yes, this is follow-up material, as it's unrelated to this PR.

garrett commented:
> There might be some UX issues left for when no mapping is found, we probably want to show something, maybe?

Yep. There should be some kind of fallback. What information is available in these cases? User & ID? Anything else?

> Or when you are in limited access and click on a root container you'll see nothing. We could map the curent user == containerid and then show something.

Yeah. I guess if there wouldn't be a way to see it in the Podman page due to limited access, we shouldn't link it? Otherwise it's just weird, right?

jelly commented:
> > There might be some UX issues left for when no mapping is found, we probably want to show something, maybe?
> 
> Yep. There should be some kind of fallback. What information is available in these cases? User & ID? Anything else?
> 
> > Or when you are in limited access and click on a root container you'll see nothing. We could map the curent user == containerid and then show something.
> 
> Yeah. I guess if there wouldn't be a way to see it in the Podman page due to limited access, we shouldn't link it? Otherwise it's just weird, right?

We could not link it yes, so if I log in as `jelle` we should only show containers for me *or* if I have privileged access `root` containers. But I think showing a tooltip with the container owner would be nice, but it does require another cockpit.spawn for `id -un -- "$cheruid"`.

martinpitt commented:
User containers all run inside the user session. I just verified that an user container which appears in the top list also appears as user@ID.service:

![image](https://user-images.githubusercontent.com/200109/170025517-be8c1854-33d5-4434-be0e-638d44f44e8c.png)

So perhaps that whole "su to different user" is just overkill? We don't have any UI to display containers other than the system and your own ones anyway. So if we'd restrict ourselves to just these two, and ignore the other ones, that might even suffice. Knowing that "it's user 1234's fault" is a good enough hint for the admin, and you can't control this kind of usage through system-level administration (like disabling or reconfiguring systemd units) anyway.

martinpitt commented on pkg/metrics/metrics.jsx:
```diff
@@ -229,6 +232,7 @@ class CurrentMetrics extends React.Component {
             netInterfacesTx: [],
             topServicesCPU: [], // [ { name, percent } ]
             topServicesMemory: [], // [ { name, bytes } ]
+            podNameMapping: {}, // { cgroup, name }
```
This is a map, so the comment should be `cgroup → name`

martinpitt commented on pkg/metrics/metrics.jsx:
```diff
@@ -391,37 +400,62 @@ class CurrentMetrics extends React.Component {
         }
 
-        function cgroupClickHandler(name, is_user, is_container) {
+        function podNameInCache(podNameMapping, uid, containerid) {
+            if (uid in podNameMapping) {
+                return containerid in podNameMapping[uid];
+            } else {
+                return false;
+            }
+        }
```
This should be `return podNameMapping[uid] && podNameMapping[uid][containerid];`. Then you'll get either `undefined` or the actual name back, which works just as well in a boolean context. And you can re-use this for `podName()` more easily, to avoid doing the lookup twice. I.e. I suggest to rename this to `getCachedPodName()`

Also, I don't see a reason to explicitly pass `podNameMapping` here -- it's always going to be `this.state.podNameMapping` anyway, isn't it? To use this, just turn this from a `function` into a `const` arrow.

jelly replied:
Nice, that's a good cleanup!

martinpitt commented on pkg/metrics/metrics.jsx:
```diff
@@ -391,37 +400,62 @@ class CurrentMetrics extends React.Component {
+        }
+
+        function podName(podNameMapping, uid, containerid) {
+            if (podNameInCache(podNameMapping, uid, containerid)) {
+                return podNameMapping[uid][containerid];
+            } else {
+                return containerid.substr(0, 12);
+            }
```
Then this is just `getCachedPodName(uid, containerid) || containerid.substr(0, 12)`.

martinpitt commented on pkg/metrics/metrics.jsx:
```diff
@@ -391,37 +400,62 @@ class CurrentMetrics extends React.Component {
         }
 
-        function cgroupRow(name, value, is_user, is_container) {
+        function cgroupRow(name, value, is_user, is_container, uid, podNameMapping) {
```
Again, if you make this a const arrow, you don't need to pass podNameMapping here at all, but just access it on this.state.

jelly replied:
This is now fixed by using `getCachedPodName` everywhere

martinpitt commented on pkg/metrics/metrics.jsx:
```diff
@@ -440,17 +474,68 @@ class CurrentMetrics extends React.Component {
+            ([key, value, is_user, is_container, userid]) => cgroupRow(key, cockpit.format_bytes(value), is_user, is_container, userid, this.state.podNameMapping)
         );
 
+        const notMappedContainers = topServicesMemory.concat(topServicesCPU).filter(([key, value, is_user, is_container, userid]) => is_container && !podNameInCache(this.state.podNameMapping, userid, key));
+        if (notMappedContainers.length !== 0) {
+            this.update_podman_name_mapping(notMappedContainers);
+        }
```
You need to take negative cache lookups into account here. I.e. if you previously failed to map an ID to a pod name (perhaps it's a docker/LXC/etc. container or custom cgroup), the cache must have an entry for that ID, mapping it to `null` (whereas "not attempted yet" would be `undefined` or even "key does not exist", which is by and large `undefined` as well). This should then *not* cause another cache refresh.

That should work fine wiht the cache building below, as that only ever updates podNameMapping instead of rewriting it.

martinpitt commented on pkg/metrics/metrics.jsx:
```diff
@@ -440,17 +474,68 @@ class CurrentMetrics extends React.Component {
+    update_podman_name_mapping = (cgroups) => {
+        const { privileged } = this.state;
+        const spawnPromise = (cmd, uid) => {
+            return new Promise((resolve, reject) => {
+                cockpit.spawn(cmd, { superuser: privileged ? "required" : null }).then(result => {
+                    resolve([result, uid]);
+                }).catch(err => reject(err));
+            });
```
Unnecessary Promise wrapper, just return cockpit.spawn(). (or more precisely, the chained promise after .then()/.catch().

martinpitt commented on pkg/metrics/metrics.jsx:
```diff
@@ -440,17 +474,68 @@ class CurrentMetrics extends React.Component {
+                return privileged ? spawnPromise(cmd, userid) : null;
+            } else {
+                if (privileged) {
+                    return spawnPromise(["sudo", "-u", "#" + userid].concat(cmd), userid);
```
This is actually rather clever, if we really want to go ahead with mapping arbitrary user IDs.

martinpitt commented on pkg/metrics/metrics.jsx:
```diff
@@ -440,17 +474,68 @@ class CurrentMetrics extends React.Component {
+                        podNameMapping[uid][container.Id] = container.Names[0];
+                    }
+                } catch (err) {
+                    console.error("podman ps outputs invalid JSON", err.toString());
```
In case that ever happens, I think this needs to set some flag (in the negative cache or separate) to entirely stop this polling, otherwise you'll poll in vain for each iteration.

martinpitt commented on pkg/metrics/metrics.jsx:
```diff
@@ -440,17 +474,68 @@ class CurrentMetrics extends React.Component {
+                    console.error("podman ps outputs invalid JSON", err.toString());
+                }
+            }
+            this.setState({ podNameMapping: Object.assign(podNameMapping, this.state.podNameMapping) });
```
OK, just FYI: In ES6 you can use the spread operator for this: `{...podNameMapping, ...this.state.podNameMapping}`.

martinpitt reviewed (changes requested):
Thanks! Happy to gmeet if you want to discuss the approach with more bandwidth.

jelly commented on pkg/metrics/metrics.jsx:
```diff
@@ -217,6 +219,7 @@ class CurrentMetrics extends React.Component {
         this.cgroupMemoryNames = [];
 
         this.state = {
+            privileged: false,
```
Ah, well we do need to keep track of it right or do we handle this cache by switching to superuser / dropping permissions reloading the page?

Took it from https://github.com/cockpit-project/cockpit/blob/main/pkg/shell/hosts.jsx#L73

martinpitt replied:
hosts.jsx looks even weirder -- nothing actually uses that state, so dead code?

Sure you need to react on a change (that's the empty setState() below), but the code can look at `superuser.allowed` directly.

jelly commented:
> User containers all run inside the user session. I just verified that an user container which appears in the top list also appears as [user@ID.service](mailto:user@ID.service):
> 
> ![image](https://user-images.githubusercontent.com/200109/170025517-be8c1854-33d5-4434-be0e-638d44f44e8c.png)
> 
> So perhaps that whole "su to different user" is just overkill? We don't have any UI to display containers other than the system and your own ones anyway. So if we'd restrict ourselves to just these two, and ignore the other ones, that might even suffice. Knowing that "it's user 1234's fault" is a good enough hint for the admin, and you can't control this kind of usage through system-level administration (like disabling or reconfiguring systemd units) anyway.

I've updated the PR and simplified it a lot. We now only show the root and logged in users containers, for root we fallback to the podman id if the user is not a superuser.

This works pretty well, the only thing missing is to squash the commits and add a test. This would be easy as it just requires waiting one render longer for the name to be resolved.

martinpitt commented on pkg/metrics/metrics.jsx:
```diff
@@ -391,37 +400,56 @@ class CurrentMetrics extends React.Component {
+        };
 
-        function cgroupClickHandler(name, is_user, is_container) {
+        function cgroupClickHandler(name, is_user, is_container, uid) {
```
I'm surprised that this works. Doesn't the `function` here introduce a new `this` context, and thus getCachedPodName() can't access this.state any more?

martinpitt commented on pkg/metrics/metrics.jsx:
```diff
@@ -391,37 +400,56 @@ class CurrentMetrics extends React.Component {
-        }
+        };
+
+        const getCachedPodName = (uid, containerid) => {
+            return this.state.podNameMapping[uid] && this.state.podNameMapping[uid][containerid];
```
Can you please drop the `{ return }` wrapper and just directly return the value?

martinpitt commented on pkg/metrics/metrics.jsx:
```diff
@@ -440,17 +468,65 @@ class CurrentMetrics extends React.Component {
 
+    /**
+     * Look up the container names using podman ps for the given cgroups. When
+     * the logged in user has administrative access we use sudo to call podman
+     * ps for that user.
```
Outdated comment

martinpitt commented on pkg/metrics/metrics.jsx:
```diff
@@ -440,17 +468,65 @@ class CurrentMetrics extends React.Component {
+            if ((userid === 0 && !superuser.allowed) && userid !== this.state.userid) {
+                return null;
+            }
+            return cockpit.spawn(["podman", "ps", "--format", "json"], { superuser: userid === 0 ? "required" : null }).then(result => {
```
Please move the .then() to the next line, it's easy to miss. Please also drop the `{ return }` wrapper.

martinpitt commented on pkg/metrics/metrics.jsx:
```diff
@@ -440,17 +468,65 @@ class CurrentMetrics extends React.Component {
+                    console.error("podman ps outputs invalid JSON", err.toString());
+                }
+            }
+            this.setState({ podNameMapping: { ...podNameMapping, ...this.state.podNameMapping } });
+        }).catch(err => {
+            this.setState({ podNameMapping: { ...podNameMapping, ...this.state.podNameMapping } });
```
DRY -- can you use `.finally()` for that?

martinpitt commented on test/verify/check-metrics:
```diff
@@ -796,6 +796,11 @@ class TestCurrentMetrics(MachineCase):
                 b.wait_in_text("#current-metrics-card-cpu", f"pod {shortid}")
             b.wait(lambda: topServiceValue(self, "Top 5 CPU services", "%", 1) > 70)
+
+            # It takes one re-render for the name lookup
+            with b.wait_timeout(300):
```
... but that sholdn't take 5 minutes surely? This is updated every 5 seconds, isn't it?

jelly replied:
Yup, so let's reduce it.

martinpitt reviewed (changes requested):
Thanks! Sorry for letting that slip. The approach looks fine to me, I just have some smaller technical quibbles. The PR is conflicty now (as you enabled this test on FCOS).

martinpitt commented:
@jelly This is almost certainly related to https://github.com/cockpit-project/cockpit/commit/b83d91e929e350 . It might be a systemd regression, but my bet is that it's an SELinux issue. Worth sending a no-test PR with `setenforce 0` to run on packit.

martinpitt commented on pkg/metrics/metrics.jsx:
```diff
@@ -229,6 +232,7 @@ class CurrentMetrics extends React.Component {
             netInterfacesTx: [],
             topServicesCPU: [], // [ { name, percent } ]
             topServicesMemory: [], // [ { name, bytes } ]
+            podNameMapping: {}, // { containerid -> name }
```
I admittedly liked this better as `cgroup`, but as it's now a two-level dict, the comment is still wrong. It's "uid → containerid → name", isn't it?

jelly replied:
It is still wrong :|, it is indeed uid => containerid => name

martinpitt commented on test/verify/check-metrics:
```diff
@@ -793,6 +793,11 @@ class TestCurrentMetrics(MachineCase):
         with b.wait_timeout(300):
             b.wait_in_text("#current-metrics-card-cpu", f"pod {shortid}")
```
This is now a race condition. It could be that the name lookup is fast enough that we don't see the short ID on the page ever. But *shrug*, it seems to work in CI, so fine for now. Just something to watch out for later.

martinpitt reviewed (approved):
Thanks! I don't have the heart to nack this just because of the wrong comment, so please go ahead! Thanks for bearing with me :grin:

martinpitt reviewed (approved):
Cheers!

martinpitt commented:
@jelly : Also, this c9s TF run has worked fine :man_shrugging:
